File: python/lib/dijkstra/graph.py
from typing import List, Optional
from .types import NodeNeighbor, GraphObject


class Node:
    def __init__(self, name: str, neighbors: List[NodeNeighbor]) -> None:
        self._name = name
        self._neighbors = neighbors

    @property
    def name(self) -> str:
        return self._name

    @property
    def neighbors(self) -> List[NodeNeighbor]:
        return self._neighbors


class Graph:
    def __init__(self, graphObject: GraphObject) -> None:
        self._nodes: List[Node] = []
        for name, neighbors in graphObject.items():
            self.add_node(Node(name, neighbors))

    @property
    def nodes(self) -> List[Node]:
        return self._nodes

    def add_node(self, node: Node) -> None:
        self._nodes.append(node)

    def find_node_by_name(self, name: str) -> Optional[Node]:
        for node in self._nodes:
            if node.name == name:
                return node
        return None

File: typescript/lib/dijkstra/graph.ts
import { type GraphObject, type NodeNeighbor } from './types';

export class Node {
  constructor(
    private readonly _name: string,
    private readonly _neighbors: NodeNeighbor[],
  ) {}

  get name(): string {
    return this._name;
  }

  get neighbors(): NodeNeighbor[] {
    return this._neighbors;
  }
}

export class Graph {
  private readonly _nodes: Node[] = [];

  constructor(graphObject: GraphObject) {
    for (const [name, neighbors] of Object.entries(graphObject)) {
      this.addNode(new Node(name, neighbors));
    }
  }

  get nodes(): Node[] {
    return this._nodes;
  }

  addNode(node: Node): void {
    this._nodes.push(node);
  }

  findNodeByName(name: string): Node | undefined {
    return this._nodes.find((node) => node.name === name);
  }
}
